File: python/doublyLinkedList_test_.py
import unittest
from DoublyLinkedList import DoublyLinkedList
from DoubleLinkedListNode import DoubleLinkedListNode

class TestProgram(unittest.TestCase):

  def test_case1(self):
    list = DoublyLinkedList()
    self.assertEqual(list.head, None)
    self.assertEqual(list.tail, None)

  def test_case2(self):
    list = DoublyLinkedList()
    node1 = DoubleLinkedListNode(1)
    list.setHead(node1)
    self.assertEqual(list.head, node1)
    self.assertEqual(list.tail, node1)

  def test_case3(self):
    list = DoublyLinkedList()
    node1 = DoubleLinkedListNode(1)
    list.setHead(node1)
    node2 = DoubleLinkedListNode(2)
    list.setHead(node2)
    self.assertEqual(list.head, node2)
    self.assertEqual(list.tail, node1)
    node3 = DoubleLinkedListNode(3)
    list.setHead(node3)
    self.assertEqual(list.head, node3)
    self.assertEqual(list.head.next, node2)

  def test_case4(self):
    list = DoublyLinkedList()
    node1 = DoubleLinkedListNode(1)
    list.setHead(node1)
    self.assertEqual(list.head, node1)
    self.assertEqual(list.tail, node1)

  def test_case5(self):
    list = DoublyLinkedList()
    node1 = DoubleLinkedListNode(1)
    node2 = DoubleLinkedListNode(2)
    node3 = DoubleLinkedListNode(3)
    node4 = DoubleLinkedListNode(4)
    list.setTail(node1)
    self.assertEqual(list.head, node1)
    self.assertEqual(list.tail, node1)
    list.setTail(node2)
    self.assertEqual(list.head, node1)
    self.assertEqual(list.tail, node2)
    list.insertBefore(node1, node3)
    self.assertEqual(list.head, node3)
    self.assertEqual(list.tail, node2)
    list.insertAfter(node2, node4)
    self.assertEqual(list.head, node3)
    self.assertEqual(list.tail, node4)

  def test_case6(self):
    list = DoublyLinkedList()
    node1 = DoubleLinkedListNode(1)
    node2 = DoubleLinkedListNode(2)
    node3 = DoubleLinkedListNode(3)
    list.setTail(node1)
    list.insertAfter(node1, node2)
    list.insertAfter(node2, node3)
    list.removeNodesWithValue(1)
    self.assertEqual(list.head, node2)
    self.assertEqual(list.tail, node3)
    self.assertEqual(list.containsNodeWithValue(1), False )
    self.assertEqual(list.containsNodeWithValue(2), True )
    self.assertEqual(list.containsNodeWithValue(3), True )


if __name__ == "__main__":
  unittest.main()

File: javascript/doublyLinkedList.test.js
const Node = require('./DoubleLinkedListNode')
const DoublyLinkedList = require('./DoublyLinkedList')

test('#1 test', ()=>{
  const list = new DoublyLinkedList()
  const node = new Node('1')
  list.setHead(node)
  expect(list.head).toStrictEqual(node)
  expect(list.tail).toStrictEqual(node)
})

test('#2 test', ()=>{
  const list = new DoublyLinkedList()
  const node1 = new Node('1')
  list.setHead(node1)
  const node2 = new Node('2')
  list.setHead(node2)
  expect(list.head).toStrictEqual(node2)
  expect(list.tail).toStrictEqual(node1)
})

test('#3 test', ()=>{
  const list = new DoublyLinkedList()
  const node1 = new Node('1')
  list.setHead(node1)
  const node2 = new Node('2')
  list.setHead(node2)
  const node3 = new Node('2')
  list.setHead(node3)

  expect(list.head).toStrictEqual(node3)
  expect(node3.next).toStrictEqual(node2)
  expect(list.tail).toStrictEqual(node1)
})

test('#4 test', ()=>{
  const list = new DoublyLinkedList()
  const node1 = new Node('1')
  list.setHead(node1)
  const node2 = new Node('2')
  list.setHead(node2)
  const node3 = new Node('3')
  list.insertBefore(node1, node3)

  expect(list.head).toStrictEqual(node2)
  expect(node1.prev).toStrictEqual(node3)
  expect(node2.next).toStrictEqual(node3)
  expect(list.tail).toStrictEqual(node1)

  list.insertBefore(node2, node1)
  expect(list.head).toStrictEqual(node1)
  expect(list.tail).toStrictEqual(node3)
  expect(list.head.next).toStrictEqual(node2)
  expect(list.tail.prev).toStrictEqual(node2)
})

test('#5 test', ()=>{
  const list = new DoublyLinkedList()
  const node1 = new Node('1')
  list.setTail(node1)
  expect(list.head).toStrictEqual(node1)
  expect(list.tail).toStrictEqual(node1)
  const node2 = new Node('2')
  list.setTail(node2)
  expect(list.head).toStrictEqual(node1)
  expect(list.tail).toStrictEqual(node2)

  const node3 = new Node('3')
  list.insertAfter(node1, node3)
  expect(list.head).toStrictEqual(node1)
  expect(list.head.next).toStrictEqual(node3)
  expect(list.tail.prev).toStrictEqual(node3)
  expect(list.tail).toStrictEqual(node2)
})

test('#6 test', ()=>{
  const list  = new DoublyLinkedList()
  const node1 = new Node('1')
  const node2 = new Node('2')
  const node3 = new Node('3')
  const node4 = new Node('4')
  list.setHead(node1)
  list.insertAfter(node1, node2)
  list.insertAfter(node2, node3)
  list.insertAtPosition(1,node4)
  expect(list.head).toStrictEqual(node4)
  expect(list.tail).toStrictEqual(node3)

  const node5 = new Node('5')
  list.insertAtPosition(2,node5)
  expect(list.head).toStrictEqual(node4)
  expect(list.head.next).toStrictEqual(node5)
  expect(list.tail).toStrictEqual(node3)
})

test('#7 test', ()=>{
  const list  = new DoublyLinkedList()
  const node1 = new Node('1')
  const node2 = new Node('2')
  const node3 = new Node('3')
  list.setHead(node1)
  list.insertAfter(node1, node2)
  list.insertAfter(node2, node3)
  expect(list.containsNodeWithValue('1')).toBe(true)
  expect(list.containsNodeWithValue('3')).toBe(true)
  expect(list.containsNodeWithValue('4')).toBe(false)

  list.removeNodesWithValue('3')
  expect(list.containsNodeWithValue('3')).toBe(false)
  expect(list.head).toStrictEqual(node1)
  expect(list.tail).toStrictEqual(node2)
})
